### PR TITLE
Improve the usability of GenericContainer

### PR DESCRIPTION
**Pull request by bsideup**
What was changed:
* store envs as Map, not List, so that one can get all envs as a map:
```
container.getEnvMap()
```

* add `withEnv` with mapper:
```
.withEnv("JAVA_OPTS", it -> it.orElse("") + " -javaagent:/foo.jar")
```

* use `READ_WRITE` bind mode if none was provided:
```
.withFileSystemBind("/foo", "/bar")
```

Also deprecated some methods from `Container` interface like `fetchDockerDaemonInfo` because we don't want to have them in our public API

### CODE REVIEW

**Comment by rnorth** [comment]
@bsideup CircleCI failure? https://circleci.com/gh/testcontainers/testcontainers-java/829
🤔

File: core/src/main/java/org/testcontainers/containers/Container.java [hunk 1]
```diff
@@ -383,6 +408,7 @@ default Integer getFirstMappedPort() {
      */
     SELF withLogConsumer(Consumer<OutputFrame> consumer);
 
+    @Deprecated
```
Comment (rnorth) [inline core/src/main/java/org/testcontainers/containers/Container.java]: Is this related to the change? Same question for deprecation `getContainerInfo()`.

I don't mind deprecating these, as it makes things cleaner, but would it be possible to have a comment indicating a recommended alternative approach?

**Comment by bsideup** [comment]
@rnorth CircleCI failed because it was using the old configuration:
https://circleci.com/gh/testcontainers/testcontainers-java/829#config/containers/0

I've added deprecation notes and entries to CHANGELOG.md :)
